llvm/lib/Target/AArch64/AArch64FrameLowering.cpp

```diff
@@ -2666,10 +2658,20 @@ void AArch64FrameLowering::processFunctionBeforeFrameFinalized(
   while (MBBI != MBB.end() && MBBI->getFlag(MachineInstr::FrameSetup))
     ++MBBI;
 
+  // Find the minimum fixed object offset already allocated.
+  int64_t MinFixedObjOffset = -getOffsetOfLocalArea();
+  for (int I = MFI.getObjectIndexBegin(); I < 0; ++I)
+    MinFixedObjOffset = std::min(MinFixedObjOffset, MFI.getObjectOffset(I));
+
+  // Ensure alignment.
+  MinFixedObjOffset -= std::abs(MinFixedObjOffset) % 16;
+
   // Create an UnwindHelp object.
+  int64_t UnwindHelpOffset = MinFixedObjOffset - 16;
   int UnwindHelpFI =
-      MFI.CreateStackObject(/*size*/8, /*alignment*/16, false);
+      MFI.CreateFixedObject(/*size*/8, UnwindHelpOffset, /*IsImmutable=*/false);
```

Comment:
We allocate this UnwindHelp object at a fixed offset so it is not affected by alignment.

This is necessary as it is accessed during unwind by the runtime at an offset relative to FP.

Reply:
Unless I am mistaken, `MinFixedObjectOffset` will always be negative. That said, this code (which I modified from X86) is only able to fix certain mis-alignments (off by 8) so I will do some more validation here to ensure I am covering all the necessary cases.

I was under the belief I need to leave the stack aligned to 16 bytes here, which may imply that it is so aligned on entry, which means I can possibly just adjust down by 16 rather than relying on an adjustment and alignment that will always result in -16.

Reply:
I believe the AArch64 ABI mandates 16-byte alignments to the stack, yes. Possibly even AArch32.